polynote-kernel/src/main/scala/polynote/data/ReprFormat.scala

```diff
@@ -0,0 +1,61 @@
+package polynote.data
+
+import scodec.bits.{BitVector, ByteVector}
+import shapeless.{Generic, HList, HNil, Lazy, ::}
+
+import scala.collection.GenTraversable
+
+trait ReprFormat[T] extends (T => ByteVector)
+
+object ReprFormat extends ReprFormat0 {
+
+  final case class instance[T](fn: T => ByteVector) extends ReprFormat[T] {
+    def apply(t: T): ByteVector = fn(t)
+  }
+
+  private val one = ByteVector.fromByte(1)
+  private val zero = ByteVector.fromByte(0)
+
+  implicit val byteFormat: ReprFormat[Byte] = instance(ByteVector.fromByte)
+  implicit val boolFormat: ReprFormat[Boolean] = instance(b => if (b) one else zero)
+  implicit val shortFormat: ReprFormat[Short] = instance(ByteVector.fromShort(_))
+  implicit val intFormat: ReprFormat[Int] = instance(ByteVector.fromInt(_))
+  implicit val longFormat: ReprFormat[Long] = instance(ByteVector.fromLong(_))
+  implicit val floatFormat: ReprFormat[Float] = instance(f => ByteVector.fromInt(java.lang.Float.floatToIntBits(f)))
+  implicit val doubleFormat: ReprFormat[Double] = instance(d => ByteVector.fromLong(java.lang.Double.doubleToLongBits(d)))
+  implicit val stringFormat: ReprFormat[String] = instance(str => ByteVector.encodeUtf8(str).fold(throw _, identity))
+  implicit val byteArrayFormat: ReprFormat[Array[Byte]] = instance(ByteVector(_))
+  implicit val byteVectorFormat: ReprFormat[ByteVector] = instance(identity)
+  implicit val bitVectorFormat: ReprFormat[BitVector] = instance(_.toByteVector)
+
+  implicit def arrayFormat[A](implicit elFormat: ReprFormat[A]): ReprFormat[Array[A]] = instance {
+    arr => arr.foldLeft(ByteVector.fromInt(arr.length)) {
+      (accum: ByteVector, next: A) => accum ++ elFormat(next)
+    }
+  }
```

Comment:
annoying that this duplication needs to exist but I guess its because Array is all special n junk